File: server/gemini.js
import { getPublicIntel } from './publicIntel.js'

const DEFAULT_GEMINI_ENDPOINT = 'https://generativelanguage.googleapis.com/v1beta'
const DEFAULT_GEMINI_MODEL = 'gemini-2.5-flash'

function trimText(value) {
  return typeof value === 'string' ? value.trim() : ''
}

function normalizeModelName(model = DEFAULT_GEMINI_MODEL) {
  return String(model).replace(/^models\//, '').trim() || DEFAULT_GEMINI_MODEL
}

function buildGeminiUrl(model, apiKey) {
  const url = new URL(`${DEFAULT_GEMINI_ENDPOINT}/models/${normalizeModelName(model)}:generateContent`)
  url.searchParams.set('key', apiKey)
  return url.toString()
}

function extractGeminiText(data) {
  const parts = data?.candidates?.[0]?.content?.parts

  if (!Array.isArray(parts)) {
    return ''
  }

  return parts
    .map((part) => trimText(part?.text))
    .filter(Boolean)
    .join('\n')
    .trim()
}

function getGeminiErrorMessage(data, fallback = '') {
  const apiMessage = trimText(data?.error?.message)
  if (apiMessage) {
    return apiMessage
  }

  const blockReason = trimText(data?.promptFeedback?.blockReason)
  if (blockReason) {
    return `Запрос заблокирован: ${blockReason}.`
  }

  return trimText(fallback) || 'unknown error'
}

function normalizeHistory(history = []) {
  return history
    .filter((item) => trimText(item?.content))
    .slice(-6)
    .map((item) => ({
      role: item.role === 'assistant' ? 'model' : 'user',
      parts: [{ text: trimText(item.content) }],
    }))
}

function dedupeStrings(values) {
  return [...new Set(values.filter(Boolean))]
}

function buildChatSystemInstruction() {
  return [
    'You are AquaMind AI Center for an Almaty smart water operations dashboard.',
    'Answer in Russian.',
    'Blend three sources of truth when relevant: local dashboard telemetry, public weather context, and public web data from Google Search grounding.',
    'Prioritize operationally relevant public factors only: weather, public works, outages, safety alerts, and major events that can affect water operations.',
    'Ignore broad history, culture, generic economics, and background facts unless they directly affect today or the next few days.',
    'Clearly distinguish between local dashboard data and public external context when that distinction matters.',
    'Do not invent incidents, regulations, outages, or sensor values.',
    'Keep the answer concise, product-grade, and operational.',
    'Use at most one short paragraph and at most three bullet points.',
    'When public web data was used, include practical implications and rely only on grounded findings.',
  ].join(' ')
}

function buildChatContextPrompt(context, alerts, knowledgeBase, publicIntel) {
  const alertText =
    alerts?.length
      ? alerts
          .slice(0, 5)
          .map((alert) => `- [${alert.severity}] ${alert.title}: ${alert.description} Действие: ${alert.action}`)
          .join('\n')
      : '- Активных нарушений по локальной телеметрии не обнаружено.'

  const knowledgeText =
    knowledgeBase?.length
      ? knowledgeBase
          .slice(0, 10)
          .map((item) => `- ${item.title}: ${item.content}`)
          .join('\n')
      : '- Локальный индекс пуст.'

  const publicSignals = publicIntel?.signals?.length
    ? publicIntel.signals.map((item) => `- ${item.label}: ${item.description}`).join('\n')
    : '- Публичные сигналы недоступны.'

  return [
    `Дата и время: ${new Date().toLocaleString('ru-RU')}`,
    '',
    'Локальный контекст дашборда:',
    `- Район: ${context.district}`,
    `- Сценарий: ${context.snapshot.scenarioLabel}`,
    `- Статус района: ${context.analysis.tone}`,
    `- Давление: ${context.snapshot.overview.pressure} bar`,
    `- Качество: ${context.snapshot.overview.quality}`,
    `- Покрытие: ${context.snapshot.overview.coverage}%`,
    `- Жалобы: ${context.snapshot.overview.complaints}`,
    `- ETA реакции: ${context.snapshot.overview.responseEta} мин`,
    '',
    'Активные локальные алерты:',
    alertText,
    '',
    'Локальный индекс и briefing:',
    knowledgeText,
    '',
    'Публичный внешний контекст:',
    publicIntel
      ? `- ${publicIntel.summary}`
      : '- Публичный внешний контекст временно недоступен.',
    publicIntel
      ? `- Пиковое внешнее окно: ${publicIntel.peakHour.label}, стресс ${publicIntel.peakHour.stressIndex}, ветер ${publicIntel.peakHour.windSpeed} км/ч, вероятность осадков ${publicIntel.peakHour.precipitationProbability}%.`
      : '',
    'Публичные сигналы:',
    publicSignals,
  ]
    .filter(Boolean)
    .join('\n')
}

function extractGroundingMetadata(data) {
  const metadata = data?.candidates?.[0]?.groundingMetadata
  const chunks = metadata?.groundingChunks || []
  const supports = metadata?.groundingSupports || []
  const queries = metadata?.webSearchQueries || []

  const webSources = chunks
    .map((chunk, index) => {
      const web = chunk?.web
      if (!web?.uri) {
        return null
      }

      const snippet =
        supports.find((item) => item?.groundingChunkIndices?.includes(index))?.segment?.text ||
        ''

      return {
        title: web.title || new URL(web.uri).hostname,
        url: web.uri,
        snippet: trimText(snippet),
        source: 'Google Search',
      }
    })
    .filter(Boolean)
    .reduce((acc, item) => {
      if (acc.some((existing) => existing.url === item.url)) {
        return acc
      }

      acc.push(item)
      return acc
    }, [])

  return {
    queries,
    webSources,
  }
}

async function fetchGemini({ apiKey, model, body }) {
  const response = await fetch(buildGeminiUrl(model, apiKey), {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(25000),
  })

  const rawText = await response.text()
  let data = null

  if (rawText) {
    try {
      data = JSON.parse(rawText)
    } catch {
      data = null
    }
  }

  if (!response.ok) {
    throw new Error(`Google AI Studio API error ${response.status}: ${getGeminiErrorMessage(data, rawText)}`)
  }

  return data
}

async function fetchFallbackChatText({ apiKey, model, prompt }) {
  const data = await fetchGemini({
    apiKey,
    model,
    body: {
      contents: [
        {
          role: 'user',
          parts: [{ text: prompt }],
        },
      ],
      generationConfig: {
        temperature: 0.25,
        maxOutputTokens: 320,
        thinkingConfig: {
          thinkingBudget: 0,
        },
      },
    },
  })

  return extractGeminiText(data)
}

export async function createGroundedChatReply({
  apiKey,
  model = DEFAULT_GEMINI_MODEL,
  context,
  alerts = [],
  knowledgeBase = [],
  history = [],
  message,
}) {
  const publicIntel = await getPublicIntel({
    district: context?.district,
    scenario: context?.activeScenario,
  })

  const data = await fetchGemini({
    apiKey,
    model,
    body: {
      systemInstruction: {
        parts: [{ text: buildChatSystemInstruction() }],
      },
      contents: [
        {
          role: 'user',
          parts: [{ text: buildChatContextPrompt(context, alerts, knowledgeBase, publicIntel) }],
        },
        ...normalizeHistory(history),
        {
          role: 'user',
          parts: [{ text: trimText(message) }],
        },
      ],
      tools: [{ googleSearch: {} }],
      generationConfig: {
        temperature: 0.35,
        maxOutputTokens: 420,
        thinkingConfig: {
          thinkingBudget: 0,
        },
      },
    },
  })

  let text = extractGeminiText(data)

  if (!text) {
    text = await fetchFallbackChatText({
      apiKey,
      model,
      prompt: [
        'Ответь кратко и по делу на русском языке.',
        'Используй локальный контекст дашборда и публичный погодный фон ниже.',
        buildChatContextPrompt(context, alerts, knowledgeBase, publicIntel),
        '',
        `Вопрос пользователя: ${trimText(message)}`,
      ].join('\n'),
    })
  }

  if (!text) {
    throw new Error(getGeminiErrorMessage(data, 'Google AI Studio вернул пустой ответ.'))
  }

  const finishReason = data?.candidates?.[0]?.finishReason
  if (finishReason === 'MAX_TOKENS' || text.length > 900) {
    const compressed = await fetchFallbackChatText({
      apiKey,
      model,
      prompt: [
        'Сожми ответ до одного короткого абзаца и трёх коротких пунктов максимум.',
        'Сохрани только то, что реально важно для оперативного решения по водоснабжению.',
        '',
        text,
      ].join('\n'),
    })

    if (compressed) {
      text = compressed
    }
  }

  const grounding = extractGroundingMetadata(data)
  const webCards = grounding.webSources.slice(0, 4).map((item) => ({
    title: item.title,
    snippet: item.snippet || item.url,
    source: item.source,
    url: item.url,
  }))

  return {
    tone: alerts[0]?.severity || context.analysis?.tone || publicIntel.tone || 'normal',
    headline: grounding.webSources.length ? 'Google AI Studio + общедоступные данные' : 'Google AI Studio ответ',
    content: text,
    highlights: dedupeStrings([
      `Локальный район: ${context.district}.`,
      `Публичный фон: ${publicIntel.current.weatherLabel}, ${publicIntel.current.temperature}°C, ветер ${publicIntel.current.windSpeed} км/ч.`,
      grounding.queries[0] ? `Веб-поиск: ${grounding.queries[0]}.` : 'Google Search подключается по мере необходимости.',
    ]),
    sources: dedupeStrings([
      'Google AI Studio',
      'Google Search',
      publicIntel.source,
      'Локальный контекст AI-центра',
    ]),
    searchResults: webCards,
    webSources: grounding.webSources.slice(0, 6),
    publicIntel,
  }
}

function buildReportSystemInstruction() {
  return [
    'You are AquaMind, an AI command analyst for an urban water supply dashboard.',
    'Return only valid JSON in Russian.',
    'The JSON keys must be: headline, summary, diagnosis, executive, engineering, citizen, recommendations, tone, confidence.',
    'recommendations must be an array of short strings.',
    'tone must be one of: normal, warning, critical.',
    'confidence must be an integer from 0 to 100.',
    'Use local dashboard context and public weather context when relevant.',
  ].join(' ')
}

function buildReportPrompt(context, mission, publicIntel) {
  return `
Контекст smart city dashboard:
- Район: ${context.district}
- Сценарий: ${context.activeScenario}
- Режим обзора: ${context.viewMode}
- Статус: ${context.analysis.tone}
- Уровень воды: ${context.snapshot.overview.waterLevel}%
- Давление: ${context.snapshot.overview.pressure} bar
- Инциденты: ${context.snapshot.overview.incidents}
- Качество воды: ${context.snapshot.overview.quality}
- Покрытие сервиса: ${context.snapshot.overview.coverage}%
- Среднее покрытие по городу: ${context.city.averageCoverage}%
- Среднее давление по городу: ${context.city.averagePressure} bar
- Всего инцидентов по городу: ${context.city.totalIncidents}

Публичный внешний контекст:
- ${publicIntel.summary}
- Пиковое внешнее окно: ${publicIntel.peakHour.label}
- Пиковый внешний стресс: ${publicIntel.peakHour.stressIndex}
- Вероятность осадков в пик: ${publicIntel.peakHour.precipitationProbability}%
- Ветер в пик: ${publicIntel.peakHour.windSpeed} км/ч

Задача:
${mission}

Сформируй:
1. diagnosis
2. executive
3. engineering
4. citizen
5. recommendations

Тон должен быть кратким, продуктовым и пригодным для реального городского штаба.
  `.trim()
}

const reportSchema = {
  type: 'object',
  properties: {
    headline: { type: 'string' },
    summary: { type: 'string' },
    diagnosis: { type: 'string' },
    executive: { type: 'string' },
    engineering: { type: 'string' },
    citizen: { type: 'string' },
    recommendations: {
      type: 'array',
      items: { type: 'string' },
    },
    tone: {
      type: 'string',
      enum: ['normal', 'warning', 'critical'],
    },
    confidence: { type: 'integer' },
  },
  required: ['headline', 'summary', 'diagnosis', 'executive', 'engineering', 'citizen', 'recommendations', 'tone', 'confidence'],
}

function parseJsonCandidate(text) {
  try {
    return JSON.parse(text)
  } catch {
    const start = text.indexOf('{')
    const end = text.lastIndexOf('}')

    if (start === -1 || end === -1 || end <= start) {
      return null
    }

    try {
      return JSON.parse(text.slice(start, end + 1))
    } catch {
      return null
    }
  }
}

export async function createStructuredReport({
  apiKey,
  model = DEFAULT_GEMINI_MODEL,
  context,
  mission,
}) {
  const publicIntel = await getPublicIntel({
    district: context?.district,
    scenario: context?.activeScenario,
  })

  const data = await fetchGemini({
    apiKey,
    model,
    body: {
      systemInstruction: {
        parts: [{ text: buildReportSystemInstruction() }],
      },
      contents: [
        {
          role: 'user',
          parts: [{ text: buildReportPrompt(context, mission, publicIntel) }],
        },
      ],
      generationConfig: {
        temperature: 0.25,
        maxOutputTokens: 900,
        thinkingConfig: {
          thinkingBudget: 0,
        },
        responseMimeType: 'application/json',
        responseJsonSchema: reportSchema,
      },
    },
  })

  const text = extractGeminiText(data)
  const parsed = parseJsonCandidate(text)

  if (!parsed || typeof parsed !== 'object') {
    throw new Error('Google AI Studio вернул невалидный JSON для отчёта.')
  }

  return {
    ...parsed,
    source: 'gemini',
    publicIntel,
  }
}

File: back/api/services/gemini.py
import json
from urllib.parse import urlencode

import requests

from .public_intel import get_public_intel

DEFAULT_GEMINI_ENDPOINT = "https://generativelanguage.googleapis.com/v1beta"
DEFAULT_GEMINI_MODEL = "gemini-2.5-flash"


def trim_text(value):
    return value.strip() if isinstance(value, str) else ""


def normalize_model_name(model=DEFAULT_GEMINI_MODEL):
    normalized = str(model or DEFAULT_GEMINI_MODEL).replace("models/", "").strip()
    return normalized or DEFAULT_GEMINI_MODEL


def build_gemini_url(model, api_key):
    return f"{DEFAULT_GEMINI_ENDPOINT}/models/{normalize_model_name(model)}:generateContent?{urlencode({'key': api_key})}"


def extract_gemini_text(data):
    parts = (((data or {}).get("candidates") or [{}])[0].get("content") or {}).get("parts")
    if not isinstance(parts, list):
        return ""
    return "\n".join(trim_text(part.get("text")) for part in parts if trim_text(part.get("text"))).strip()


def get_gemini_error_message(data, fallback=""):
    api_message = trim_text(((data or {}).get("error") or {}).get("message"))
    if api_message:
        return api_message
    block_reason = trim_text(((data or {}).get("promptFeedback") or {}).get("blockReason"))
    if block_reason:
        return f"Запрос заблокирован: {block_reason}."
    return trim_text(fallback) or "unknown error"


def normalize_history(history=None):
    normalized = []
    for item in (history or [])[-6:]:
        content = trim_text(item.get("content"))
        if not content:
            continue
        normalized.append(
            {
                "role": "model" if item.get("role") == "assistant" else "user",
                "parts": [{"text": content}],
            }
        )
    return normalized


def dedupe_strings(values):
    unique = []
    for value in values:
        if value and value not in unique:
            unique.append(value)
    return unique


def build_chat_system_instruction():
    return " ".join(
        [
            "You are AquaMind AI Center for an Almaty smart water operations dashboard.",
            "Answer in Russian.",
            "Blend three sources of truth when relevant: local dashboard telemetry, public weather context, and public web data from Google Search grounding.",
            "Prioritize operationally relevant public factors only: weather, public works, outages, safety alerts, and major events that can affect water operations.",
            "Clearly distinguish between local dashboard data and public external context when that distinction matters.",
            "Do not invent incidents, regulations, outages, or sensor values.",
            "Keep the answer concise, product-grade, and operational.",
            "Use at most one short paragraph and at most three bullet points.",
        ]
    )


def build_chat_context_prompt(context, alerts, knowledge_base, public_intel):
    alerts = alerts or []
    knowledge_base = knowledge_base or []
    alert_text = (
        "\n".join(
            f"- [{alert.get('severity')}] {alert.get('title')}: {alert.get('description')} Действие: {alert.get('action')}"
            for alert in alerts[:5]
        )
        if alerts
        else "- Активных нарушений по локальной телеметрии не обнаружено."
    )
    knowledge_text = (
        "\n".join(f"- {item.get('title')}: {item.get('content')}" for item in knowledge_base[:10])
        if knowledge_base
        else "- Локальный индекс пуст."
    )
    public_signals = (
        "\n".join(f"- {item.get('label')}: {item.get('description')}" for item in public_intel.get("signals", []))
        if public_intel and public_intel.get("signals")
        else "- Публичные сигналы недоступны."
    )
    snapshot = context.get("snapshot") or {}
    overview = snapshot.get("overview") or {}
    analysis = context.get("analysis") or {}

    return "\n".join(
        [
            "Локальный контекст дашборда:",
            f"- Район: {context.get('district')}",
            f"- Сценарий: {snapshot.get('scenarioLabel')}",
            f"- Статус района: {analysis.get('tone')}",
            f"- Давление: {overview.get('pressure')} bar",
            f"- Качество: {overview.get('quality')}",
            f"- Покрытие: {overview.get('coverage')}%",
            f"- Жалобы: {overview.get('complaints')}",
            f"- ETA реакции: {overview.get('responseEta')} мин",
            "",
            "Активные локальные алерты:",
            alert_text,
            "",
            "Локальный индекс и briefing:",
            knowledge_text,
            "",
            "Публичный внешний контекст:",
            f"- {public_intel.get('summary')}" if public_intel else "- Публичный внешний контекст временно недоступен.",
            (
                f"- Пиковое внешнее окно: {public_intel['peakHour']['label']}, стресс {public_intel['peakHour']['stressIndex']}, "
                f"ветер {public_intel['peakHour']['windSpeed']} км/ч, вероятность осадков {public_intel['peakHour']['precipitationProbability']}%."
            )
            if public_intel
            else "",
            "Публичные сигналы:",
            public_signals,
        ]
    )


def extract_grounding_metadata(data):
    metadata = (((data or {}).get("candidates") or [{}])[0].get("groundingMetadata") or {})
    chunks = metadata.get("groundingChunks") or []
    supports = metadata.get("groundingSupports") or []
    queries = metadata.get("webSearchQueries") or []
    web_sources = []
    for index, chunk in enumerate(chunks):
        web = chunk.get("web") or {}
        if not web.get("uri"):
            continue
        snippet = ""
        for support in supports:
            if index in (support.get("groundingChunkIndices") or []):
                snippet = trim_text(((support.get("segment") or {}).get("text")))
                break
        item = {
            "title": web.get("title") or web.get("uri"),
            "url": web.get("uri"),
            "snippet": snippet,
            "source": "Google Search",
        }
        if not any(existing["url"] == item["url"] for existing in web_sources):
            web_sources.append(item)
    return {"queries": queries, "webSources": web_sources}


def fetch_gemini(api_key, model, body):
    response = requests.post(
        build_gemini_url(model, api_key),
        headers={"content-type": "application/json"},
        json=body,
        timeout=25,
    )
    raw_text = response.text
    try:
        data = response.json() if raw_text else None
    except ValueError:
        data = None
    if not response.ok:
        raise RuntimeError(f"Google AI Studio API error {response.status_code}: {get_gemini_error_message(data, raw_text)}")
    return data


def fetch_fallback_chat_text(api_key, model, prompt):
    data = fetch_gemini(
        api_key,
        model,
        {
            "contents": [{"role": "user", "parts": [{"text": prompt}]}],
            "generationConfig": {
                "temperature": 0.25,
                "maxOutputTokens": 320,
                "thinkingConfig": {"thinkingBudget": 0},
            },
        },
    )
    return extract_gemini_text(data)


def create_grounded_chat_reply(api_key, context, message, alerts=None, knowledge_base=None, history=None, model=DEFAULT_GEMINI_MODEL):
    public_intel = get_public_intel(context.get("district"), context.get("activeScenario"))
    data = fetch_gemini(
        api_key,
        model,
        {
            "systemInstruction": {"parts": [{"text": build_chat_system_instruction()}]},
            "contents": [
                {"role": "user", "parts": [{"text": build_chat_context_prompt(context, alerts, knowledge_base, public_intel)}]},
                *normalize_history(history),
                {"role": "user", "parts": [{"text": trim_text(message)}]},
            ],
            "tools": [{"googleSearch": {}}],
            "generationConfig": {
                "temperature": 0.35,
                "maxOutputTokens": 420,
                "thinkingConfig": {"thinkingBudget": 0},
            },
        },
    )
    text = extract_gemini_text(data)
    if not text:
        text = fetch_fallback_chat_text(
            api_key,
            model,
            "\n".join(
                [
                    "Ответь кратко и по делу на русском языке.",
                    "Используй локальный контекст дашборда и публичный погодный фон ниже.",
                    build_chat_context_prompt(context, alerts, knowledge_base, public_intel),
                    "",
                    f"Вопрос пользователя: {trim_text(message)}",
                ]
            ),
        )
    if not text:
        raise RuntimeError(get_gemini_error_message(data, "Google AI Studio вернул пустой ответ."))

    finish_reason = (((data or {}).get("candidates") or [{}])[0].get("finishReason"))
    if finish_reason == "MAX_TOKENS" or len(text) > 900:
        compressed = fetch_fallback_chat_text(
            api_key,
            model,
            "\n".join(
                [
                    "Сожми ответ до одного короткого абзаца и трёх коротких пунктов максимум.",
                    "Сохрани только то, что реально важно для оперативного решения по водоснабжению.",
                    "",
                    text,
                ]
            ),
        )
        if compressed:
            text = compressed

    grounding = extract_grounding_metadata(data)
    web_cards = [
        {
            "title": item["title"],
            "snippet": item["snippet"] or item["url"],
            "source": item["source"],
            "url": item["url"],
        }
        for item in grounding["webSources"][:4]
    ]
    return {
        "tone": (alerts or [{}])[0].get("severity") or (context.get("analysis") or {}).get("tone") or public_intel.get("tone") or "normal",
        "headline": "Google AI Studio + общедоступные данные" if grounding["webSources"] else "Google AI Studio ответ",
        "content": text,
        "highlights": dedupe_strings(
            [
                f"Локальный район: {context.get('district')}.",
                f"Публичный фон: {public_intel['current']['weatherLabel']}, {public_intel['current']['temperature']}°C, ветер {public_intel['current']['windSpeed']} км/ч.",
                f"Веб-поиск: {grounding['queries'][0]}." if grounding["queries"] else "Google Search подключается по мере необходимости.",
            ]
        ),
        "sources": dedupe_strings(["Google AI Studio", "Google Search", public_intel["source"], "Локальный контекст AI-центра"]),
        "searchResults": web_cards,
        "webSources": grounding["webSources"][:6],
        "publicIntel": public_intel,
    }


def build_report_system_instruction():
    return " ".join(
        [
            "You are AquaMind, an AI command analyst for an urban water supply dashboard.",
            "Return only valid JSON in Russian.",
            "The JSON keys must be: headline, summary, diagnosis, executive, engineering, citizen, recommendations, tone, confidence.",
            "recommendations must be an array of short strings.",
            "tone must be one of: normal, warning, critical.",
            "confidence must be an integer from 0 to 100.",
            "Use local dashboard context and public weather context when relevant.",
        ]
    )


def build_report_prompt(context, mission, public_intel):
    snapshot = context.get("snapshot") or {}
    overview = snapshot.get("overview") or {}
    city = context.get("city") or {}
    analysis = context.get("analysis") or {}
    return (
        f"Контекст smart city dashboard:\n"
        f"- Район: {context.get('district')}\n"
        f"- Сценарий: {context.get('activeScenario')}\n"
        f"- Режим обзора: {context.get('viewMode')}\n"
        f"- Статус: {analysis.get('tone')}\n"
        f"- Уровень воды: {overview.get('waterLevel')}%\n"
        f"- Давление: {overview.get('pressure')} bar\n"
        f"- Инциденты: {overview.get('incidents')}\n"
        f"- Качество воды: {overview.get('quality')}\n"
        f"- Покрытие сервиса: {overview.get('coverage')}%\n"
        f"- Среднее покрытие по городу: {city.get('averageCoverage')}%\n"
        f"- Среднее давление по городу: {city.get('averagePressure')} bar\n"
        f"- Всего инцидентов по городу: {city.get('totalIncidents')}\n\n"
        f"Публичный внешний контекст:\n"
        f"- {public_intel['summary']}\n"
        f"- Пиковое внешнее окно: {public_intel['peakHour']['label']}\n"
        f"- Пиковый внешний стресс: {public_intel['peakHour']['stressIndex']}\n"
        f"- Вероятность осадков в пик: {public_intel['peakHour']['precipitationProbability']}%\n"
        f"- Ветер в пик: {public_intel['peakHour']['windSpeed']} км/ч\n\n"
        f"Задача:\n{mission}\n\n"
        f"Сформируй:\n1. diagnosis\n2. executive\n3. engineering\n4. citizen\n5. recommendations\n\n"
        f"Тон должен быть кратким, продуктовым и пригодным для реального городского штаба."
    )


REPORT_SCHEMA = {
    "type": "object",
    "properties": {
        "headline": {"type": "string"},
        "summary": {"type": "string"},
        "diagnosis": {"type": "string"},
        "executive": {"type": "string"},
        "engineering": {"type": "string"},
        "citizen": {"type": "string"},
        "recommendations": {"type": "array", "items": {"type": "string"}},
        "tone": {"type": "string", "enum": ["normal", "warning", "critical"]},
        "confidence": {"type": "integer"},
    },
    "required": ["headline", "summary", "diagnosis", "executive", "engineering", "citizen", "recommendations", "tone", "confidence"],
}


def parse_json_candidate(text):
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        start = text.find("{")
        end = text.rfind("}")
        if start == -1 or end == -1 or end <= start:
            return None
        try:
            return json.loads(text[start : end + 1])
        except json.JSONDecodeError:
            return None


def create_structured_report(api_key, context, mission, model=DEFAULT_GEMINI_MODEL):
    public_intel = get_public_intel(context.get("district"), context.get("activeScenario"))
    data = fetch_gemini(
        api_key,
        model,
        {
            "systemInstruction": {"parts": [{"text": build_report_system_instruction()}]},
            "contents": [{"role": "user", "parts": [{"text": build_report_prompt(context, mission, public_intel)}]}],
            "generationConfig": {
                "temperature": 0.25,
                "maxOutputTokens": 900,
                "thinkingConfig": {"thinkingBudget": 0},
                "responseMimeType": "application/json",
                "responseJsonSchema": REPORT_SCHEMA,
            },
        },
    )
    text = extract_gemini_text(data)
    parsed = parse_json_candidate(text)
    if not isinstance(parsed, dict):
        raise RuntimeError("Google AI Studio вернул невалидный JSON для отчёта.")
    return {**parsed, "source": "gemini", "publicIntel": public_intel}
